include/substrait/type/Type.h

```diff
@@ -0,0 +1,701 @@
+};
+
+class Struct : public TypeBase<TypeKind::kStruct> {
+ public:
+  explicit Struct(std::vector<TypePtr> types, bool nullable = false)
+      : TypeBase<TypeKind::kStruct>(nullable), children_(std::move(types)) {}
+
+  [[nodiscard]] std::string signature() const override;
+
+  [[nodiscard]] const std::vector<TypePtr>& children() const {
+    return children_;
+  }
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const std::vector<TypePtr> children_;
+};
+
+class Map : public TypeBase<TypeKind::kMap> {
+ public:
+  Map(TypePtr keyType, TypePtr valueType, bool nullable = false)
+      : TypeBase<TypeKind::kMap>(nullable),
+        keyType_(std::move(keyType)),
+        valueType_(std::move(valueType)) {}
+
+  [[nodiscard]] const TypePtr& keyType() const {
+    return keyType_;
+  }
+
+  [[nodiscard]] const TypePtr& valueType() const {
+    return valueType_;
+  }
+
+  [[nodiscard]] std::string signature() const override;
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const TypePtr keyType_;
+  const TypePtr valueType_;
+};
+
+/// ParameterizedType represent a type in
+class ParameterizedTypeBase : public ParameterizedType {
+ public:
+  explicit ParameterizedTypeBase(bool nullable = false)
+      : ParameterizedType(nullable) {}
+};
+
+class UsedDefinedType : public ParameterizedTypeBase {
+ public:
+  UsedDefinedType(std::string value, bool nullable)
+      : ParameterizedTypeBase(nullable), value_(std::move(value)) {}
+
+  [[nodiscard]] const std::string& value() const {
+    return value_;
+  }
+
+  [[nodiscard]] TypeKind kind() const override {
+    return TypeKind::kUserDefined;
+  }
+
+  [[nodiscard]] std::string signature() const override {
+    return TypeTraits<TypeKind::kUserDefined>::signature;
+  }
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  /// raw string of wildcard type.
+  const std::string value_;
+};
+
+/// A string literal type can present the 'any1'.
+class StringLiteral : public ParameterizedTypeBase {
+ public:
+  explicit StringLiteral(std::string value)
+      : ParameterizedTypeBase(false), value_(std::move(value)) {}
+
+  [[nodiscard]] std::string signature() const override {
+    return value_;
+  }
+
+  [[nodiscard]] TypeKind kind() const override {
+    return TypeKind::KIND_NOT_SET;
+  }
+
+  [[nodiscard]] const std::string& value() const {
+    return value_;
+  }
+
+  [[nodiscard]] bool isWildcard() const override {
+    return value_.find("any") == 0 || value_ == "T";
+  }
+
+  [[nodiscard]] bool isMatch(
+      const std::shared_ptr<const ParameterizedType>& type) const override;
+
+ private:
+  const std::string value_;
+};
+
+using StringLiteralPtr = std::shared_ptr<const StringLiteral>;
+
+class ParameterizedDecimal : public ParameterizedTypeBase {
```

Comment:
How is `ParameterizedDecimal` different from `Decimal`?

Reply:
Have a look at this [ParameterizedDecimal can take ](https://github.com/substrait-io/substrait/blob/main/proto/substrait/parameterized_types.proto#L84),  ParametrizedDecimal can take string or integer as type parameter